proposals/views/proposal_views.py

```diff
@@ -541,36 +542,57 @@ def get_context_data(self, **kwargs):
         return context
 
 
-class ProposalAsPdf(LoginRequiredMixin, PDFTemplateResponseMixin,
-                    generic.DetailView):
+class ProposalAsPdf(
+        LoginRequiredMixin,
+        generic.DetailView,
+        PDFTemplateResponseMixin,
+):
     model = Proposal
-    template_name = 'proposals/proposal_pdf.html'
-
     # The PDF mixin generates a filename with this factory
     filename_factory = FilenameFactory('Proposal')
 
+    def get(self, request, *args, **kwargs):
+        # First, check if we should use a pregenerated pdf, if we have one
+        proposal = self.get_object()
+        if proposal.use_canonical_pdf():
+            if proposal.pdf:
+                return FileResponse(
+                    proposal.pdf,
+                    filename=self.get_pdf_filename(),
+                    as_attachment=self.pdf_save_as,
+                )
+        # Else, continue with generation
+        return super().get(request, *args, **kwargs)
+
+    def get_object(self):
+        """If we already have an object set, use that.
+        This can happen if this view is being called from generate_pdf()
+        rather than by Django."""
+        if not hasattr(self, "object"):
+            self.object = super().get_object()
+        return self.object
+
+    def get_template_names(self):
+        """Determine the correct PDf template for given proposal"""
+        self.template_name = 'proposals/proposal_pdf.html'
```

Comment:
This first line seems redundant?

Reply:
This vestigial appendage has been dealt with